wen-restart/src/last_voted_fork_slots_aggregate.rs

```diff
@@ -3,69 +3,108 @@ use {
-    solana_runtime::epoch_stakes::EpochStakes,
-    solana_sdk::{clock::Slot, hash::Hash, pubkey::Pubkey},
+    solana_runtime::bank::Bank,
+    solana_sdk::{
+        clock::{Epoch, Slot},
+        hash::Hash,
+        pubkey::Pubkey,
+    },
     std::{
-        collections::{HashMap, HashSet},
+        collections::{BTreeSet, HashMap, HashSet},
         str::FromStr,
+        sync::Arc,
     },
 };
 
+// If at least 1/3 of the stake has voted for a slot in next Epoch, we think
+// the cluster's clock is in sync and everyone will enter the new Epoch soon.
+// So we require that we have >80% stake in the new Epoch to exit.
+const EPOCH_CONSIDERED_FOR_EXIT_THRESHOLD: f64 = 1f64 / 3f64;
+
+#[derive(Debug, Clone, PartialEq)]
+pub(crate) struct LastVotedForkSlotsEpochInfo {
+    pub epoch: Epoch,
+    pub total_stake: u64,
+    // Percentage (0 to 100) of total stake of active peers in this epoch,
+    // no matter they voted for a slot in this epoch or not.
+    pub voted_percent: f64,
+    // Percentage (0 to 100) of total stake of active peers which has voted for
+    // a slot in this epoch.
+    pub voted_for_this_epoch_percent: f64,
+}
+
 pub(crate) struct LastVotedForkSlotsAggregate {
-    root_slot: Slot,
-    repair_threshold: f64,
-    // TODO(wen): using local root's EpochStakes, need to fix if crossing Epoch boundary.
-    epoch_stakes: EpochStakes,
+    // Map each peer pubkey to the epoch of its last vote.
+    node_to_last_vote_epoch_map: HashMap<Pubkey, Epoch>,
+    epoch_info_vec: Vec<LastVotedForkSlotsEpochInfo>,
     last_voted_fork_slots: HashMap<Pubkey, RestartLastVotedForkSlots>,
-    slots_stake_map: HashMap<Slot, u64>,
-    active_peers: HashSet<Pubkey>,
-    slots_to_repair: HashSet<Slot>,
     my_pubkey: Pubkey,
+    repair_threshold: f64,
+    root_bank: Arc<Bank>,
+    slots_stake_map: HashMap<Slot, u64>,
+    slots_to_repair: BTreeSet<Slot>,
 }
 
 #[derive(Clone, Debug, PartialEq)]
 pub struct LastVotedForkSlotsFinalResult {
     pub slots_stake_map: HashMap<Slot, u64>,
-    pub total_active_stake: u64,
+    pub epoch_info_vec: Vec<LastVotedForkSlotsEpochInfo>,
 }
 
 impl LastVotedForkSlotsAggregate {
     pub(crate) fn new(
-        root_slot: Slot,
+        root_bank: Arc<Bank>,
         repair_threshold: f64,
-        epoch_stakes: &EpochStakes,
         last_voted_fork_slots: &Vec<Slot>,
         my_pubkey: &Pubkey,
     ) -> Self {
-        let mut active_peers = HashSet::new();
-        let sender_stake = Self::validator_stake(epoch_stakes, my_pubkey);
-        active_peers.insert(*my_pubkey);
         let mut slots_stake_map = HashMap::new();
+        let root_slot = root_bank.slot();
+        let root_epoch = root_bank.epoch();
         for slot in last_voted_fork_slots {
             if slot >= &root_slot {
-                slots_stake_map.insert(*slot, sender_stake);
+                if let Some(sender_stake) = root_bank.epoch_node_id_to_stake(root_epoch, my_pubkey)
```

Comment:
Should this stake be based on the epoch `slot` is in rather than the stake in the `root_epoch`? 

Latest vote slot could be bigger than root epoch

Also nit: `last_voted_fork_slots` -> `my_last_voted_fork_slots`

Reply:
Ah, good catch, changed. Also added a testcase.

renamed.